Why Globalize?

rxaviers opened the pull request:
Include "Why Globalize?" section on README.

jzaefferer commented on README.md:
```diff
@@ -76,6 +77,29 @@ providing an API to convert user-entered number and date strings - in their
+<a name="why_globalize"></a>
+### Why Globalize?
+
+Globalize has initially been created to fullfil jQuery UI (datepicker and
```
Well, initially Microsoft wanted to provide a JS globalization library, it had nothing to do with jQuery UI. If we keep this sentence, the grammar needs to be improved, e.g., "has ben created to fullfil the requirements of jQuery UI" or "to provide globalization support for jQuery UI"

rxaviers replied:
Removed

jzaefferer commented on README.md:
```diff
@@ -76,6 +77,29 @@ providing an API to convert user-entered number and date strings - in their
+
+Our design goals are (a) to leverage the official CLDR JSON data, (b) allow
+users to load as much or as little data as they need, (c) avoid duplicating data
+if using multiple i18n libraries that leverage CLDR, and (d) run in browsers or
```
"run in browsers and node.js."

jzaefferer commented on README.md:
```diff
@@ -76,6 +77,29 @@ providing an API to convert user-entered number and date strings - in their
+formatting and parsing, currency formatting, message formatting (ICU message
+format pattern), and plural support.
+
+Our design goals are (a) to leverage the official CLDR JSON data, (b) allow
```
I'd also say "Make globalization as easy to use as jQuery"

jzaefferer commented on README.md:
```diff
@@ -76,6 +77,29 @@ providing an API to convert user-entered number and date strings - in their
+
+- Based on CLDR data (largest and most extensive i18n repository).
+- No bugs due to outdated i18n content (it has no hard coded i18n data).
+- Consistency. Same results are obtained across browsers.
```
Add "and on the server"

rxaviers replied:
@jzaefferer,

There are two comparisons that are important to emphasize. One, getting consistent results between different browsers. Two, getting consistent results between a browser and the server.

After adding "and on the server" to the sentence ("Same results are obtained across browsers and on the server"), the second emphasis is privileged.

The below, although repetitive, makes it clear. Can we improve it?

"- Consistent. Same results are obtained across browsers. Same results are obtained between browsers and the server."

jzaefferer commented on README.md:
```diff
@@ -76,6 +77,29 @@ providing an API to convert user-entered number and date strings - in their
+- Based on CLDR data (largest and most extensive i18n repository).
+- No bugs due to outdated i18n content (it has no hard coded i18n data).
+- Consistency. Same results are obtained across browsers.
+- Allow different locale instances running simultaneously.
```
"Support using multiple locales at the same time"

jzaefferer commented on README.md:
```diff
@@ -76,6 +77,29 @@ providing an API to convert user-entered number and date strings - in their
+- Consistency. Same results are obtained across browsers.
+- Allow different locale instances running simultaneously.
+
+For an overview of JavaScript globalization libraries, check out [this
```
Rewriting myself: "For alternative libraries and more, check out this [JavaScript globalization overview]."

jzaefferer commented on README.md:
```diff
@@ -12,7 +12,8 @@ Node.js module.
   - [Not accepting 0.x fixes anymore](#0.x-fixes)
 - [About Globalize](#about)
-  - [Why globalization?](#why)
+  - [Why globalization?](#why_globalization)
```
Would be good to add something along the lines of "Even if you're only dealing with the English locale, you are still likely to need globalization, to display something better than "Edited 1 minutes ago"."

rxaviers replied:
What do you think of https://github.com/jquery/globalize/pull/392/files#diff-04c6e90faac2675aa89e2176d2eec7d8R80

jzaefferer commented on README.md:
```diff
@@ -76,6 +77,28 @@ providing an API to convert user-entered number and date strings - in their
+- Consistency. Same results are obtained across browsers and on the server.
+- Support using multiple locales simultaneously.
+
+For alternative libraries and more, check out [this
```
You didn't update the link text, I suggested "check out this [JavaScript globalization overview](...)"

rxaviers commented:
Thanks for your review @jzaefferer , it was of a great help.

jzaefferer commented on README.md:
```diff
@@ -76,6 +77,32 @@ providing an API to convert user-entered number and date strings - in their
 own format - into actual numbers and dates, and conversely, to format numbers
 and dates into that string format.
 
+Even if the application deals with Enligh loacle only, it may still likely need
```
"deals only with the English locale, it may still need"

"may" and "likely" don't go well together, one is enough to indicate uncertainty.

rxaviers replied:
updated

jzaefferer commented on README.md:
```diff
@@ -76,6 +77,32 @@ providing an API to convert user-entered number and date strings - in their
 
+Even if the application deals with Enligh loacle only, it may still likely need
+globalization to format programming language bytes into human understandable
+language and vice-versa in a effective and reasonable way.
```
If you prefer this, please also add my example. That was heavily inspired by Alex Sexton's intro to messageformat.js. The example makes it much more obvious and uses several aspects of Globalize that are not trivial to implement ad-hoc.

jzaefferer commented on README.md:
```diff
@@ -76,6 +77,32 @@ providing an API to convert user-entered number and date strings - in their
+- Consistency. Same results are obtained across browsers and on the server.
+- Support using multiple locales simultaneously.
+
+For alternative libraries and more, check out [this JavaScript globalization
```
I didn't put "this" into the link text...

rxaviers commented:
@agcolom @kswedberg @arthurvr, please if you have any feedback, feel free to add your comments. Thanks

scottgonzalez commented on README.md:
```diff
@@ -58,7 +59,7 @@ Are you looking for 0.x docs? Find them [here](https://github.com/jquery/globali
 ## About Globalize
 
-<a name="why"></a>
+<a name="why_globalization"></a>
```
Why aren't you just using the auto-generated header links?

rxaviers replied:
It has been inherited from old README. But, could be updated to use the auto-generated ones.

rxaviers replied:
Since, we'll eventually migrate README into api.globalizejs.com, I'm not too concerned with this for now.

scottgonzalez replied:
But even then you should be using the auto-generated header links.

rxaviers replied:
Ok, let's track that via https://github.com/jquery/globalize/issues/393

scottgonzalez commented on README.md:
```diff
@@ -76,10 +77,37 @@ providing an API to convert user-entered number and date strings - in their
+parsing, currency formatting, message formatting (ICU message format pattern),
+and plural support.
+
+Our design goals are (a) to leverage the official CLDR JSON data, (b) allow
```
Why not split this out into an actual list?

kswedberg replied:
:+1: Use a list like you do with the strengths below. I'd also recommend using "Design Goals" and "Strengths of Using Globalize" as headings.

arschmitz replied:
This threw me off, it does render as an actual bulleted list. https://github.com/jquery/globalize/blob/why-globalize/README.md#why-globalize

rxaviers replied:
Updated as items list.

rxaviers replied:
Also updated titles as @kswedberg suggests.

scottgonzalez replied:
It might make sense to have the list and then have a short explanation of each point afterward. For example, someone new to globalization isn't going to know why "leverage the official CLDR JSON data" is something they should even care about.

kswedberg commented on README.md:
```diff
@@ -76,10 +77,37 @@ providing an API to convert user-entered number and date strings - in their
 and dates into that string format.
 
+Even if the application deals only with the English locale, it may still need
+globalization to format programming language bytes into human understandable
```
"human understandable" should be hyphenated: "human-understandable"

rxaviers replied:
Fixed thanks

kswedberg commented on README.md:
```diff
@@ -76,10 +77,37 @@ providing an API to convert user-entered number and date strings - in their
 
+Even if the application deals only with the English locale, it may still need
+globalization to format programming language bytes into human understandable
+language and vice-versa in a effective and reasonable way. For example, to
```
"a effective" -> "an effective"

rxaviers replied:
Fixed thanks

kswedberg commented on README.md:
```diff
@@ -76,10 +77,37 @@ providing an API to convert user-entered number and date strings - in their
+Some of the strengths of using Globalize are:
+
+- Based on CLDR data (largest and most extensive i18n repository).
+- No bugs due to outdated i18n content (it has no hard coded i18n data).
```
"hard coded" -> "hard-coded"

rxaviers replied:
Fixed thanks

rxaviers replied:
@scottgonzalez, is there any remaining point we need to explain?

kswedberg commented on README.md:
```diff
@@ -76,10 +77,37 @@ providing an API to convert user-entered number and date strings - in their
+- No bugs due to outdated i18n content (it has no hard coded i18n data).
+- Consistency. Same results are obtained across browsers and on the server.
+- Support using multiple locales simultaneously.
+
```
Make the above list items (more) parallel:

* Based on… (leave as is)
* Consistent: Same results…
* Avoids bugs caused by outdated…
* Supports using…

rxaviers replied:
"Avoids" and "Supports" are in the singular form. Should the verbs of the above list ("Leverage", "Allow", "Avoid", "Run", and "Make") all follow the singular form as well?

rxaviers replied:
About making the list more parallel, what do you think of having all items as imperatives?

- Use the largest and most extensive i18n repository. Globalize is based on CLDR
  data.
- Avoid bugs caused by outdated i18n content. Globalize has no hard-coded i18n
  data. Up-to-date CLDR data can be used directly as soon as it's released.
- Get same results across different browsers. Get same results between browsers
  and the server.
- Use multiple locales simultaneously.

scottgonzalez commented on README.md:
```diff
@@ -76,10 +77,37 @@ providing an API to convert user-entered number and date strings - in their
+Some of the strengths of using Globalize are:
+
+- Based on CLDR data (largest and most extensive i18n repository).
+- No bugs due to outdated i18n content (it has no hard coded i18n data).
```
These are perfect examples of explaining the benefits of each point above.

rxaviers commented:
Thanks @scottgonzalez and @kswedberg for your review.

kswedberg commented:
No problem. Ping me any time, @rxaviers !

rxaviers commented on README.md:
```diff
@@ -76,10 +77,40 @@ providing an API to convert user-entered number and date strings - in their
+
+- Based on CLDR data (largest and most extensive i18n repository).
+- No bugs due to outdated i18n content (it has no hard-coded i18n data).
+- Consistency. Same results are obtained across browsers and on the server.
```
@jzaefferer,

There are two comparisons that are important to emphasize. One, getting consistent results between different browsers. Two, getting consistent results between a browser and the server.

After adding "and on the server" to the sentence ("Same results are obtained across browsers and on the server"), the second emphasis is privileged.

The below, although repetitive, makes it clear. Can we improve it?

"- Consistent. Same results are obtained across browsers. Same results are obtained between browsers and the server."

rxaviers commented:
I've added "Follow Unicode Technical Standard #35" as another design goal.

kswedberg commented on README.md:
```diff
@@ -76,10 +77,44 @@ providing an API to convert user-entered number and date strings - in their
+- Get consistent results across different browsers. Get consistent results
+  between browsers and the server.
+- Use multiple locales simultaneously.
+
```
@rxaviers : I really like the improvements you made to this list. Great job. Using the imperative here feels weird to me, though, since the list is referring to "strengths." One easy change would be to keep everything as is except change the verbs to declarative:

* Uses…
* Avoids…
* Gets… Gets…
* Uses…

rxaviers commented:
I think we have too many bullet points now. Looking at https://github.com/jquery/globalize/blob/why-globalize/README.md#why-globalize is confusing.

Less items with higher impact would be better.

Design goals and Strengths are somewhat related. Can we have "Strengths" and "Weaknesses" only?

scottgonzalez commented:
> I think we have too many bullet points now. Looking at https://github.com/jquery/globalize/blob/why-globalize/README.md#why-globalize is confusing.
> 
> Less items with higher impact would be better.

When I said:

> It might make sense to have the list and then have a short explanation of each point afterward. For example, someone new to globalization isn't going to know why "leverage the official CLDR JSON data" is something they should even care about.

I didn't intend for there to be two bulleted lists. The short explanation of each point afterward would be short paragraphs, providing an explanation that is beneficial to new and experienced developers.

For example, instead of:

Use the largest and most extensive i18n repository. Globalize is based on CLDR data.

You could have something like:

Globalize is based on the Unicode Consortium's Common Locale Data Repository (CLDR), the largest and most extensive standard repository of locale data available. CLDR is constantly updated and is used by many large applications and operating systems, so you'll always have access to the most accurate and up-to-date locale data.

You could also provide information about the negative aspect of not using CLDR to drive home why this is important when looking at other projects. The important part is that you're actually explaining the strengths and instilling confidence in Globalize.

rxaviers commented:
Following @kswedberg's suggestion, I'm using declarative form verbs "(Globalize) Leverages..., (Globalize) allows..., etc".

Following @scottgonzalez's  suggestion, I'm using full paragraphs to explain the strengths.

I have updated the "Design Goals" by grouping some similar goals into the same bulleted item. I've also added some missing goals.

rxaviers commented:
I've added a paragraph about Ecma-402.

rxaviers commented:
I'm interpreting the 4 days of silence as this being good to land. Thanks everyone for your reviews.

scottgonzalez commented:
Just got around to reading the updates. This looks great!

rxaviers commented:
Excellent, thanks.
